bit-sliced range index for int, long, float, double, dictionarized SV columns

richardstartin opened the pull request:
## Description

Implements new range indexing for SV columns. Relates to #7437.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [x] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
Introduce new range index implementation which can be enabled by setting `"rangeIndexVersion": 2` to `"tableIndexConfig"`. This will result in faster range queries particularly for timestamp and latency columns, and similar or better performance for floating point columns.

<!-- If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text.
-->
## Documentation
<!-- If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document
-->

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java:
```diff
@@ -304,6 +306,13 @@ public void setReadMode(ReadMode readMode) {
   public Map<String, Map<String, String>> getColumnProperties() {
     return _columnProperties;
   }
+  public int getRangeIndexVersion() {
```
@Jackie-Jiang it's not 100% clear to me if this is the right place to be configuring this, please let me know your thoughts.

Jackie-Jiang replied:
`IndexLoadingConfig` is not exposed to the user, but initialized from the table config. We can put this index version into the `IndexingConfig` which is a sub-config under the `TableConfig` so that user can specify the range index version they want to generate.

richardstartin replied:
I originally put this in `IndexingConfig` but unfortunately the table config was null in some unit tests, which looks like an error. I'll move it back.

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitSlicedRangeIndexCreator.java:
```diff
@@ -0,0 +1,168 @@
+    if (dataType == FLOAT) {
+      return 1L << 22;
+    }
+    return Long.MAX_VALUE;
```
This should be the maximum dictionary code

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitSlicedRangeIndexCreator.java:
```diff
@@ -0,0 +1,168 @@
+    if (minValue instanceof Number && maxValue instanceof Number) {
+      if (dataType == DOUBLE || dataType == FLOAT) {
+        // if the range were zero, we wouldn't be creating a range index
+        // min and max should be finite (but this needs to be verified)
```
What guarantees do we have that min and max values are finite?

richardstartin replied:
This no longer matters.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitSlicedRangeIndexCreator.java:
```diff
@@ -0,0 +1,168 @@
+  }
+
+  @Override
+  public void add(int[] dictIds, int length) {
```
throw `UnsupportedOperationException` to prevent callers from calling this method

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitSlicedRangeIndexCreator.java:
```diff
@@ -0,0 +1,168 @@
+        return ((Number) maxValue).longValue() - ((Number) minValue).longValue();
+      }
+    }
+    if (dataType == DOUBLE) {
```
Putting some comments on how it works would be good

richardstartin replied:
This is a WIP and will change, but I will be sure to document the mapping clearly when it is ready.

richardstartin replied:
This is encapsulated in the class `FPOrdering` now.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/column/PhysicalColumnIndexContainer.java:
```diff
@@ -175,7 +176,15 @@ public PhysicalColumnIndexContainer(SegmentDirectory.Reader segmentReader, Colum
+        } else if (version == 1) {
+          _rangeIndex = new BitSlicedRangeIndexReader(buffer, metadata);
+        } else {
+          _rangeIndex = null;
```
Suggest throwing exception when getting an file of unknown version

richardstartin replied:
What about logging that we didn't build the index? I could imagine evolving this in the future and not throwing would keep this forward compatible in mixed version clusters with a more graceful failure mode (just don't use an index we don't currently understand).

Jackie-Jiang replied:
Yeah, we should also log when we find the index is configured but not possible to be built (e.g. for string columns).
A more graceful failure mode sounds good to me. We might want to log a warning on not loading the unknown index

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitSlicedRangeIndexReader.java:
```diff
@@ -0,0 +1,171 @@
+  private final double _scale;
+  private volatile SoftReference<RangeBitmap> _rangeBitmapRef;
+
+  public BitSlicedRangeIndexReader(PinotDataBuffer dataBuffer, ColumnMetadata metadata) {
```
Dictionary encoded column is not handled in the reader

richardstartin replied:
It is supported by querying for `int` ranges. For dictionary encoded columns, the `RangeBitmap` will have been built over dictionary ids, as `BitSlicedRangeIndexCreator` implements `DictionaryBasedInvertedIndexCreator`. 

Are you concerned that double/float columns might have a dictionary (I would not expect this configuration)?

Jackie-Jiang replied:
Double/float columns can have dictionary. Sometimes these columns can have low cardinality, and dictionary-encoding gives better performance.

richardstartin replied:
I think this is now correct, the reader doesn't care what the data type is and no longer depends on `ColumnMetadata`. It expects to be queried with the right types (ints for dictionarized or raw int columns, long, float, or double for LONG, FLOAT, or DOUBLE raw columns. 

If the caller can't/shouldn't be trusted to provide column-type appropriate arguments, I can put type checking/conversions in.

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/IndexLoadingConfig.java:
```diff
@@ -49,6 +49,7 @@
 public class IndexLoadingConfig {
   private static final int DEFAULT_REALTIME_AVG_MULTI_VALUE_COUNT = 2;
   public static final String DEFAULT_TIER_BACKEND = "local";
+  private static final int DEFAULT_RANGE_INDEX_VERSION = 1;
```
+1 on putting one as the default version. Once all pinot components are deployed, user can start generating the new index version

richardstartin replied:
This has moved to `IndexingConfig`

Jackie-Jiang reviewed (commented):
Looks good in general. Good job!

kishoreg commented:
> Regarding the config (`IndexingConfig`), do you think it makes sense to keep two separate lists for old value range index and new bit-sliced range index? Users might want to mix them since the new one doesn't support MV. Or we can automatically fall back to the old one for MV. Both way are okay to me.

lets not add another list. This is a better implementation. We can enhance fieldconfig to allow users to pick one impl vs other in short term but my take is slicedbitmap will be better in all cases and we wont need bucketed

richardstartin commented:
This now handles floating point numbers correctly so is in what I consider to be its final state (subject to review comments).

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitSlicedRangeIndexReader.java:
```diff
@@ -0,0 +1,150 @@
+    return lte.toMutableRoaringBitmap();
+  }
+
+  private RangeBitmap getRangeBitmap() {
+    SoftReference<RangeBitmap> ref = _rangeBitmapRef;
+    RangeBitmap rangeBitmap;
+    if (ref == null) {
+      rangeBitmap = mapRangeBitmap();
+      UPDATER.compareAndSet(this, null, new SoftReference<>(rangeBitmap));
+    } else {
+      rangeBitmap = ref.get();
+      if (rangeBitmap == null) {
+        // it's been cleared so reinitialize
+        rangeBitmap = mapRangeBitmap();
+        UPDATER.compareAndSet(this, ref, new SoftReference<>(rangeBitmap));
+      }
+    }
+    return rangeBitmap;
+  }
+
+  private RangeBitmap mapRangeBitmap() {
+    // note that this is a very cheap operation, no deserialization is required
+    ByteBuffer buffer = _dataBuffer.toDirectByteBuffer(_offset, (int) (_dataBuffer.size() - _offset));
+    return RangeBitmap.map(buffer);
+  }
```
@npawar I imagine you would like this changed so there's no caching here?

richardstartin replied:
Removed by https://github.com/apache/pinot/pull/7454/commits/04c1cfbb1fa6433ec41a96f991c66253d06dfa1c

npawar replied:
Yes, thank you! This will help with bringing in the acquire-release semantics

richardstartin commented:
Another flaky test: `HelixExternalViewBasedQueryQuotaManagerTest.testNoBrokerResource`

richardstartin commented:
Here are some numbers for this change. The benchmark creates data of a given type and a given distribution, computes the deciles of the data, and the queries for a range corresponding to the nth decile. So `_decile=0` means querying for a range from the minimum value in the segment to the largest value smaller than 90% of the segment. `_decile=5` means start at the median and match 10% of rows.

For `INT`, `FLOAT`, and `LONG` query time is consistently faster, as is index build time. Because of the lossless encoding used, `DOUBLE` is slower and is comparable to the existing range index. Several things could be done about this

1. Consider a lossy conversion, I don't like this approach and the decision would need to be made before releasing this, otherwise lossy conversion would require a new range index version for double columns
2. The data structure in `RoaringBitmap` could be made roughly 2x faster by evaluating both ranges at once. Currently we pass over the `RangeBitmap` twice, once for the upper bound, once for the lower bound, and then intersect them. This could all be done in one pass in the future.
3. The data structure could be evolved to use another base for faster queries for `DOUBLE` columns, at the expense of using more space. This would be an easy migration if necessary since the serialisation format in `RoaringBitmap` is forward compatible.

I recommend not doing anything about `DOUBLE` for now, and adopt this format, on the basis that option 2 is available.

```
Benchmark                     (_dataType)  (_decile)  (_numDocs)                          (_scenario)  (_seed)  Mode  Cnt       Score        Error  Units
BenchmarkRangeIndex.queryV1           INT          5     1000000                          NORMAL(0,1)       42  avgt    5    6724.390 ±    341.587  us/op
BenchmarkRangeIndex.queryV1           INT          5     1000000                NORMAL(10000000,1000)       42  avgt    5    1679.114 ±     58.021  us/op
BenchmarkRangeIndex.queryV1           INT          5     1000000                          EXP(0.0001)       42  avgt    5    1656.115 ±     92.468  us/op
BenchmarkRangeIndex.queryV1           INT          5     1000000                             EXP(0.5)       42  avgt    5    2755.567 ±    178.396  us/op
BenchmarkRangeIndex.queryV1           INT          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1664.481 ±     25.965  us/op
BenchmarkRangeIndex.queryV1           INT          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    1833.423 ±     65.662  us/op
BenchmarkRangeIndex.queryV1           INT          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1604.613 ±     36.730  us/op
BenchmarkRangeIndex.queryV1           INT          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    1693.864 ±     11.346  us/op
BenchmarkRangeIndex.queryV1          LONG          5     1000000                          NORMAL(0,1)       42  avgt    5    7325.071 ±    326.691  us/op
BenchmarkRangeIndex.queryV1          LONG          5     1000000                NORMAL(10000000,1000)       42  avgt    5    1638.851 ±     13.261  us/op
BenchmarkRangeIndex.queryV1          LONG          5     1000000                          EXP(0.0001)       42  avgt    5    1655.873 ±     71.568  us/op
BenchmarkRangeIndex.queryV1          LONG          5     1000000                             EXP(0.5)       42  avgt    5    2751.560 ±    264.398  us/op
BenchmarkRangeIndex.queryV1          LONG          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1729.673 ±     42.050  us/op
BenchmarkRangeIndex.queryV1          LONG          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    1832.178 ±     25.334  us/op
BenchmarkRangeIndex.queryV1          LONG          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1630.938 ±     19.844  us/op
BenchmarkRangeIndex.queryV1          LONG          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    1838.097 ±     22.097  us/op
BenchmarkRangeIndex.queryV1         FLOAT          5     1000000                          NORMAL(0,1)       42  avgt    5    1671.051 ±     34.408  us/op
BenchmarkRangeIndex.queryV1         FLOAT          5     1000000                NORMAL(10000000,1000)       42  avgt    5    1660.597 ±     39.827  us/op
BenchmarkRangeIndex.queryV1         FLOAT          5     1000000                          EXP(0.0001)       42  avgt    5    1748.752 ±    104.994  us/op
BenchmarkRangeIndex.queryV1         FLOAT          5     1000000                             EXP(0.5)       42  avgt    5    1653.907 ±     32.989  us/op
BenchmarkRangeIndex.queryV1         FLOAT          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1742.404 ±    153.677  us/op
BenchmarkRangeIndex.queryV1         FLOAT          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    8941.210 ±    362.685  us/op
BenchmarkRangeIndex.queryV1         FLOAT          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1694.232 ±     52.786  us/op
BenchmarkRangeIndex.queryV1         FLOAT          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    8708.284 ±    203.975  us/op
BenchmarkRangeIndex.queryV1        DOUBLE          5     1000000                          NORMAL(0,1)       42  avgt    5    1741.052 ±     31.400  us/op
BenchmarkRangeIndex.queryV1        DOUBLE          5     1000000                NORMAL(10000000,1000)       42  avgt    5    1806.929 ±    107.558  us/op
BenchmarkRangeIndex.queryV1        DOUBLE          5     1000000                          EXP(0.0001)       42  avgt    5    1777.536 ±     72.958  us/op
BenchmarkRangeIndex.queryV1        DOUBLE          5     1000000                             EXP(0.5)       42  avgt    5    1781.936 ±     16.409  us/op
BenchmarkRangeIndex.queryV1        DOUBLE          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1717.138 ±     21.024  us/op
BenchmarkRangeIndex.queryV1        DOUBLE          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    1826.421 ±    113.665  us/op
BenchmarkRangeIndex.queryV1        DOUBLE          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1776.522 ±     46.617  us/op
BenchmarkRangeIndex.queryV1        DOUBLE          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    1733.831 ±     22.657  us/op
BenchmarkRangeIndex.queryV2           INT          5     1000000                          NORMAL(0,1)       42  avgt    5     167.473 ±      6.449  us/op
BenchmarkRangeIndex.queryV2           INT          5     1000000                NORMAL(10000000,1000)       42  avgt    5     607.574 ±      2.534  us/op
BenchmarkRangeIndex.queryV2           INT          5     1000000                          EXP(0.0001)       42  avgt    5     761.198 ±      9.089  us/op
BenchmarkRangeIndex.queryV2           INT          5     1000000                             EXP(0.5)       42  avgt    5     357.203 ±      2.710  us/op
BenchmarkRangeIndex.queryV2           INT          5     1000000              UNIFORM(0,100000000000)       42  avgt    5     832.929 ±     17.764  us/op
BenchmarkRangeIndex.queryV2           INT          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5     309.349 ±      2.834  us/op
BenchmarkRangeIndex.queryV2           INT          5     1000000              UNIFORM(0,100000000000)       42  avgt    5     835.358 ±     16.465  us/op
BenchmarkRangeIndex.queryV2           INT          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5     308.966 ±      3.106  us/op
BenchmarkRangeIndex.queryV2          LONG          5     1000000                          NORMAL(0,1)       42  avgt    5     160.306 ±      9.662  us/op
BenchmarkRangeIndex.queryV2          LONG          5     1000000                NORMAL(10000000,1000)       42  avgt    5     607.199 ±     11.274  us/op
BenchmarkRangeIndex.queryV2          LONG          5     1000000                          EXP(0.0001)       42  avgt    5     761.519 ±      7.005  us/op
BenchmarkRangeIndex.queryV2          LONG          5     1000000                             EXP(0.5)       42  avgt    5     339.800 ±      3.553  us/op
BenchmarkRangeIndex.queryV2          LONG          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1686.816 ±    112.211  us/op
BenchmarkRangeIndex.queryV2          LONG          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5     309.759 ±      3.846  us/op
BenchmarkRangeIndex.queryV2          LONG          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1627.157 ±     37.182  us/op
BenchmarkRangeIndex.queryV2          LONG          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5     310.557 ±      8.902  us/op
BenchmarkRangeIndex.queryV2         FLOAT          5     1000000                          NORMAL(0,1)       42  avgt    5     926.658 ±     35.276  us/op
BenchmarkRangeIndex.queryV2         FLOAT          5     1000000                NORMAL(10000000,1000)       42  avgt    5     820.031 ±     33.360  us/op
BenchmarkRangeIndex.queryV2         FLOAT          5     1000000                          EXP(0.0001)       42  avgt    5    1362.855 ±     22.333  us/op
BenchmarkRangeIndex.queryV2         FLOAT          5     1000000                             EXP(0.5)       42  avgt    5     738.290 ±     18.907  us/op
BenchmarkRangeIndex.queryV2         FLOAT          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1331.245 ±     40.306  us/op
BenchmarkRangeIndex.queryV2         FLOAT          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5      24.028 ±      0.687  us/op
BenchmarkRangeIndex.queryV2         FLOAT          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    1328.662 ±     29.253  us/op
BenchmarkRangeIndex.queryV2         FLOAT          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5      24.437 ±      0.335  us/op
BenchmarkRangeIndex.queryV2        DOUBLE          5     1000000                          NORMAL(0,1)       42  avgt    5    1811.274 ±     47.818  us/op
BenchmarkRangeIndex.queryV2        DOUBLE          5     1000000                NORMAL(10000000,1000)       42  avgt    5    2006.414 ±     43.702  us/op
BenchmarkRangeIndex.queryV2        DOUBLE          5     1000000                          EXP(0.0001)       42  avgt    5    2635.751 ±     61.159  us/op
BenchmarkRangeIndex.queryV2        DOUBLE          5     1000000                             EXP(0.5)       42  avgt    5    1797.613 ±     99.108  us/op
BenchmarkRangeIndex.queryV2        DOUBLE          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    2491.316 ±    106.713  us/op
BenchmarkRangeIndex.queryV2        DOUBLE          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    1049.657 ±     26.725  us/op
BenchmarkRangeIndex.queryV2        DOUBLE          5     1000000              UNIFORM(0,100000000000)       42  avgt    5    2616.917 ±    258.422  us/op
BenchmarkRangeIndex.queryV2        DOUBLE          5     1000000  UNIFORM(100000000000, 100000000100)       42  avgt    5    1049.279 ±     13.951  us/op
BenchmarkRangeIndex.createV1          INT        N/A     1000000                          NORMAL(0,1)       42    ss    5  113182.901 ±  67834.528  us/op
BenchmarkRangeIndex.createV1          INT        N/A     1000000                NORMAL(10000000,1000)       42    ss    5  375338.881 ±   6212.163  us/op
BenchmarkRangeIndex.createV1          INT        N/A     1000000                          EXP(0.0001)       42    ss    5  436650.473 ±   7519.528  us/op
BenchmarkRangeIndex.createV1          INT        N/A     1000000                             EXP(0.5)       42    ss    5  137547.093 ±   4648.473  us/op
BenchmarkRangeIndex.createV1          INT        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  546988.945 ±  22817.827  us/op
BenchmarkRangeIndex.createV1          INT        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5  287844.486 ± 104651.014  us/op
BenchmarkRangeIndex.createV1          INT        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  538891.335 ±  64592.288  us/op
BenchmarkRangeIndex.createV1          INT        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5  292665.860 ±  39843.334  us/op
BenchmarkRangeIndex.createV1         LONG        N/A     1000000                          NORMAL(0,1)       42    ss    5  113853.468 ±  10754.627  us/op
BenchmarkRangeIndex.createV1         LONG        N/A     1000000                NORMAL(10000000,1000)       42    ss    5  386002.879 ±  13724.519  us/op
BenchmarkRangeIndex.createV1         LONG        N/A     1000000                          EXP(0.0001)       42    ss    5  442085.193 ±  14979.299  us/op
BenchmarkRangeIndex.createV1         LONG        N/A     1000000                             EXP(0.5)       42    ss    5  146690.531 ±   9511.060  us/op
BenchmarkRangeIndex.createV1         LONG        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  574195.880 ±   7166.069  us/op
BenchmarkRangeIndex.createV1         LONG        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5  283517.626 ±   5951.661  us/op
BenchmarkRangeIndex.createV1         LONG        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  554756.767 ±  12630.176  us/op
BenchmarkRangeIndex.createV1         LONG        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5  285294.799 ±   7490.806  us/op
BenchmarkRangeIndex.createV1        FLOAT        N/A     1000000                          NORMAL(0,1)       42    ss    5  475590.610 ±  17834.880  us/op
BenchmarkRangeIndex.createV1        FLOAT        N/A     1000000                NORMAL(10000000,1000)       42    ss    5  350757.276 ±  28081.516  us/op
BenchmarkRangeIndex.createV1        FLOAT        N/A     1000000                          EXP(0.0001)       42    ss    5  482675.137 ±  52474.766  us/op
BenchmarkRangeIndex.createV1        FLOAT        N/A     1000000                             EXP(0.5)       42    ss    5  477357.549 ±  25480.646  us/op
BenchmarkRangeIndex.createV1        FLOAT        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  476962.173 ±  20946.250  us/op
BenchmarkRangeIndex.createV1        FLOAT        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5   66718.894 ±  12363.014  us/op
BenchmarkRangeIndex.createV1        FLOAT        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  483928.163 ±  46663.945  us/op
BenchmarkRangeIndex.createV1        FLOAT        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5   69877.679 ±  17961.117  us/op
BenchmarkRangeIndex.createV1       DOUBLE        N/A     1000000                          NORMAL(0,1)       42    ss    5  478620.142 ±  23063.686  us/op
BenchmarkRangeIndex.createV1       DOUBLE        N/A     1000000                NORMAL(10000000,1000)       42    ss    5  489118.331 ±  23570.345  us/op
BenchmarkRangeIndex.createV1       DOUBLE        N/A     1000000                          EXP(0.0001)       42    ss    5  494501.923 ±  49732.437  us/op
BenchmarkRangeIndex.createV1       DOUBLE        N/A     1000000                             EXP(0.5)       42    ss    5  503061.507 ± 103399.593  us/op
BenchmarkRangeIndex.createV1       DOUBLE        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  489518.989 ±  29034.824  us/op
BenchmarkRangeIndex.createV1       DOUBLE        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5  488868.489 ±  25696.651  us/op
BenchmarkRangeIndex.createV1       DOUBLE        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  495284.426 ±  26534.459  us/op
BenchmarkRangeIndex.createV1       DOUBLE        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5  493155.289 ±  41312.905  us/op
BenchmarkRangeIndex.createV2          INT        N/A     1000000                          NORMAL(0,1)       42    ss    5   31557.939 ±   1621.705  us/op
BenchmarkRangeIndex.createV2          INT        N/A     1000000                NORMAL(10000000,1000)       42    ss    5  153848.542 ±  21056.818  us/op
BenchmarkRangeIndex.createV2          INT        N/A     1000000                          EXP(0.0001)       42    ss    5  251817.559 ±   9385.519  us/op
BenchmarkRangeIndex.createV2          INT        N/A     1000000                             EXP(0.5)       42    ss    5   37842.500 ±   2152.924  us/op
BenchmarkRangeIndex.createV2          INT        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  476068.705 ±   8204.144  us/op
BenchmarkRangeIndex.createV2          INT        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5   59718.807 ±   3406.474  us/op
BenchmarkRangeIndex.createV2          INT        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  409972.865 ±  10681.709  us/op
BenchmarkRangeIndex.createV2          INT        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5   65081.968 ±   2253.144  us/op
BenchmarkRangeIndex.createV2         LONG        N/A     1000000                          NORMAL(0,1)       42    ss    5   31451.596 ±   2165.520  us/op
BenchmarkRangeIndex.createV2         LONG        N/A     1000000                NORMAL(10000000,1000)       42    ss    5  151057.040 ±  13465.482  us/op
BenchmarkRangeIndex.createV2         LONG        N/A     1000000                          EXP(0.0001)       42    ss    5  250432.339 ±  11397.610  us/op
BenchmarkRangeIndex.createV2         LONG        N/A     1000000                             EXP(0.5)       42    ss    5   37703.155 ±   1805.054  us/op
BenchmarkRangeIndex.createV2         LONG        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  600135.713 ±  50377.187  us/op
BenchmarkRangeIndex.createV2         LONG        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5   67510.670 ±  20257.985  us/op
BenchmarkRangeIndex.createV2         LONG        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  592899.314 ±  10187.439  us/op
BenchmarkRangeIndex.createV2         LONG        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5   66523.598 ±  19555.141  us/op
BenchmarkRangeIndex.createV2        FLOAT        N/A     1000000                          NORMAL(0,1)       42    ss    5  407109.354 ±   5819.103  us/op
BenchmarkRangeIndex.createV2        FLOAT        N/A     1000000                NORMAL(10000000,1000)       42    ss    5  232501.197 ±  20092.900  us/op
BenchmarkRangeIndex.createV2        FLOAT        N/A     1000000                          EXP(0.0001)       42    ss    5  385567.492 ±   5394.826  us/op
BenchmarkRangeIndex.createV2        FLOAT        N/A     1000000                             EXP(0.5)       42    ss    5  454067.172 ±   5904.919  us/op
BenchmarkRangeIndex.createV2        FLOAT        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  421185.434 ±   9261.012  us/op
BenchmarkRangeIndex.createV2        FLOAT        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5   76458.581 ±   3884.131  us/op
BenchmarkRangeIndex.createV2        FLOAT        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  419558.409 ±   7258.182  us/op
BenchmarkRangeIndex.createV2        FLOAT        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5   76546.710 ±   1750.497  us/op
BenchmarkRangeIndex.createV2       DOUBLE        N/A     1000000                          NORMAL(0,1)       42    ss    5  972463.978 ±  26785.075  us/op
BenchmarkRangeIndex.createV2       DOUBLE        N/A     1000000                NORMAL(10000000,1000)       42    ss    5  674071.565 ±  36243.218  us/op
BenchmarkRangeIndex.createV2       DOUBLE        N/A     1000000                          EXP(0.0001)       42    ss    5  861501.424 ±   4628.596  us/op
BenchmarkRangeIndex.createV2       DOUBLE        N/A     1000000                             EXP(0.5)       42    ss    5  915558.381 ±  16495.181  us/op
BenchmarkRangeIndex.createV2       DOUBLE        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  871729.294 ±  10802.879  us/op
BenchmarkRangeIndex.createV2       DOUBLE        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5  458504.991 ±   7541.249  us/op
BenchmarkRangeIndex.createV2       DOUBLE        N/A     1000000              UNIFORM(0,100000000000)       42    ss    5  860718.749 ±  24117.249  us/op
BenchmarkRangeIndex.createV2       DOUBLE        N/A     1000000  UNIFORM(100000000000, 100000000100)       42    ss    5  466381.706 ±  56695.081  us/op
```

Jackie-Jiang commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/invertedindex/RangeIndexHandler.java:
```diff
@@ -48,13 +52,18 @@
     _segmentWriter = segmentWriter;
     _columnsToAddIdx = new HashSet<>(indexLoadingConfig.getRangeIndexColumns());
+    // guard against null table config
+    _rangeIndexVersion = Optional.ofNullable(indexLoadingConfig.getTableConfig())
```
We can add the `_rangeIndexVersion` as a field in the `IndexLoadingConfig`, and allow overriding it with table config. That's the way we handle other index related configs

Jackie-Jiang reviewed (commented):
The algorithm of mapping floating point value to integer range is very smart, and will not lose precision!

mqliang commented on pinot-segment-local/src/test/java/org/apache/pinot/segment/local/segment/index/creator/BitSlicedIndexCreatorTest.java:
```diff
@@ -0,0 +1,487 @@
+  public void setUp()
+      throws IOException {
+    FileUtils.forceMkdir(INDEX_DIR);
+  }
```
Add corresponding `tearDown()` function which delete the INDEX_DIR?
```
FileUtils.deleteQuietly(INDEX_DIR);
```

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/FPOrdering.java:
```diff
@@ -0,0 +1,77 @@
+   * @return an ordinal
+   */
+  public static long ordinalOf(float value) {
+    if (value == Double.NEGATIVE_INFINITY) {
+      return 0;
+    }
+    if (value == Double.POSITIVE_INFINITY || Double.isNaN(value)) {
+      return 0xFFFFFFFF;
+    }
```
Good catch, this was accidental

richardstartin commented on pinot-segment-local/src/test/java/org/apache/pinot/segment/local/utils/FPOrderingTest.java:
```diff
@@ -0,0 +1,133 @@
+
+  @Test
+  public void testNaN() {
+    assertEquals(0xFFFFFFFFFFFFFFFFL, FPOrdering.ordinalOf(Double.NaN));
```
testng is quirky ;)

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/FPOrdering.java:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * Maps doubles to longs with the same total (unsigned) order
+   * NEGATIVE_INFINITY is considered less than or equal to all values.
+   * NaN and POSITIVE_INFINITY are considered greater than or equal to all values.
```
No reason, except it's necessary to choose one. I can change it.

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitSlicedRangeIndexCreator.java:
```diff
@@ -0,0 +1,142 @@
+    if (metadata.hasDictionary()) {
+      return metadata.getCardinality();
+    }
+    FieldSpec.DataType dataType = metadata.getDataType();
```
Ideally this would be failing a test, I can provoke it in a unit test of course.

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/BitSlicedRangeIndexReader.java:
```diff
@@ -0,0 +1,120 @@
+    _min = dataBuffer.getLong(offset);
+    offset += Long.BYTES;
+
+    this._offset = offset;
```
Let's ban it with Checkstyle so we don't need to talk about it.

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/inv/BitSlicedRangeIndexCreator.java:
```diff
@@ -0,0 +1,142 @@
+import org.apache.pinot.spi.data.FieldSpec;
+import org.roaringbitmap.RangeBitmap;
+
+import static org.apache.pinot.segment.spi.V1Constants.Indexes.BITMAP_RANGE_INDEX_FILE_EXTENSION;
```
Shouldn't there be a Checkstyle rule to ban this if it's a problem?

richardstartin replied:
There are lots of static imports (including this one which is already used elsewhere) and code review is the least scalable and least consistent way to enforce code style. I created this PR to define and enforce static import style: https://github.com/apache/pinot/pull/7483

richardstartin commented:
@Jackie-Jiang I have addressed your comments here: https://github.com/apache/pinot/pull/7454/commits/2e6c4307ceb3412cdfecdd5a71a7e6c04117261c but have squashed the commits before that.

richardstartin commented on pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/FPOrdering.java:
```diff
@@ -0,0 +1,77 @@
+    // need negatives to come before positives
+    if ((bits & Long.MIN_VALUE) == Long.MIN_VALUE) {
+      // conflate 0/-0, or reverse order of negatives
+      bits = bits == Long.MIN_VALUE ? Long.MIN_VALUE : ~bits;
```
It doesn't really matter because they are identical for the purposes of comparison. I.e. x < -0.0 and x < 0.0 are identical statements.

Jackie-Jiang reviewed (approved):
LGTM! Thanks for adding the detailed benchmark results and explanation. The result is very promising!
I added a small auto-reformat commit to follow the [pinot code style](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#intellij). Haven't figured out a way to do it automatically.

richardstartin commented:
> I added a small auto-reformat commit to follow the pinot code style. Haven't figured out a way to do it automatically.

Thanks. I think we can write checkstyle rules for these changes easily.
